rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,359 @@
+a user has to obtain a user certificate with `route_to_app` for the TLS routing
+handshake, and the backend validates the identity then routes to the appropriate
+app service based on this certificate.
+
+#### Implementation - MCP server handler with JSON-RPC
+
+MCP server handler will be placed in App service's connection handler. The MCP
+server handler will receive a raw connection once the connection is authorized
+and confirmed to be an MCP app. It is expected [the MCP protocol in JSON-RPC
+format](https://modelcontextprotocol.io/specification/2025-03-26) will be
+transferred within this connection.
+
+Since using `stdio`, a `exec.Command` defined by the app definition is executed
+to start an MCP server, for each authorized incoming client connection. Then
+stdin and stdout are proxied between the client connection and the launched MCP
+server. Stderr from the launched MCP server will be logged at TRACE level in
+Teleport logs.
+
+Incoming bytes from the client do get the following treatments before they are
+forwarded to the launched MCP server:
+- Parsed into MCP/JSON-RPC messages. 
+- If the message is a `tools/call` request, use the access-checker to confirm
+  whether the identity has access to the requested tool.
+  - If the `tools/call` is access denied, make a text result detailing this
+    error and send it back to the client connection. Do not forward the request to
+    the launched MCP server in this case.
+- Record audit events for eligible messages.
+
+The launched MCP server should be stopped once the connection ends.
+
+#### Implementation - "tsh mcp" commands
+
+A suite of `tsh mcp` commands are added on the client side:
+- `tsh mcp ls`: lists MCP server apps, with slightly different columns from `tsh
+  app ls`.
+- `tsh mcp connect`: command to be called by the AI tool to launch an MCP server
+  through Teleport. User cert with `route_to_app` is automatically obtained if
+  the app hasn't been logged in yet. stdin and stdout of this process are
+  proxied to Teleport backend. Any local tsh logs are sent to stderr.
+- `tsh mcp config`: shows a sample JSON configuration that launches the MCP
+  server with `tsh mcp connect` for Claude Desktop. (Most AI tools use the same
+  JSON format as well.)
+- `tsh mcp db`: command to be called by the AI tool to launch a local MCP server
+  that provides a database query tool through Teleport Database access. This
+  feature is outside this RFD but shares the same `tsh mcp` root.
+
+#### Compatibility with App access
+
+Since the MCP servers run as apps, here’s a list of some existing app
+functionalities and how they align with it:
+- Web UI: MCP servers will be listed as Application resources. Similar to TCP
+  apps, MCP servers cannot be "launched" through Web UI, and clicking "Connect"
+  button opens a dialog with `tsh` instructions.
+- VNet: not supported.
+- Per-session MFA: not supported.
+- Access Request: supported.
+- Dynamic registration: supported.
```

Comment:
This requires extra thought. As defined, MCP servers are arbitrary commands, which escalates "create app resource" permission to "RCE on all app servers".

To make it less of a foot gun we may add extra opt-in for this, especially if we want to make the entire functionality available by default.

Reply:
We should definitely require a different "resource" permission on app server resources that represent MCP servers. So check on on `KindApp` but on `KindMCPServer` or something. Would also make it easier if/when we eventually move this to its own service.

Reply:
I've added a comment here as Roman suggested that `KindMCPServer` will be explicitly required. note that default editor does not have rule for `KindApp`, we can do the same for `KindMCPServer`